Src/VimCore/CommonOperations.fs

```diff
@@ -446,8 +446,11 @@ type internal CommonOperations
 
         /// Move the caret right.  Don't go off the end of the line
         let moveRight () =
-            if x.CaretPoint.Position < x.CaretLine.End.Position then
-                let point = SnapshotPointUtil.AddOne x.CaretPoint
+            let point =
+                x.CaretPoint
+                |> SnapshotPointUtil.AddOne
+                |> (fun y -> if Char.IsSurrogate (y.GetChar()) then SnapshotPointUtil.AddOne y else y)

```

Comment:
This code will throw in the case where the caret is positioned on the last character of a file.  This occurs because `GetChar` throws when on the end of the `ITextSnapshot`.  

Also I think we need to change this to look for the Low Surrogate.  Otherwise moving right to the start of a surrogate pair would cause it to jump to the middle.